### ISSUE TITLE
Gdk-Message: 16:00:20.069: Lost connection to Wayland compositor.

### ISSUE THREAD

**Issue by bionade24** [xournalpp/xournalpp#3707]
**Affects versions :**
 - OS: Latest Arch Linux
 - (Linux only) Desktop environment: Wayland, Sway 1.6.1,  Wlroots 0.14.1
 - libgtk 3.24.30
 - Xournal++ 1.1.0
 - Archlinux Repos

**Describe the bug**
Xournal++ terminates (no segfault!)  randomly. Before terminating, this output is written: `Gdk-Message: 16:00:20.069: Lost connection to Wayland compositor.`

**To Reproduce**
Open files, work an edit until it occurs.

**Expected behavior**
Two things:

1. Don't get such a connection loss to the compositor and work fine as all other apps do.
2. Don't just produce emergencysaves on crashes, but also in situations like this one. Or just save normally. Currently, since this isn't a real crash, xournal++ doesn't make a emergencysave.xopp

**Comment by rolandlo** [comment]
This [GIMP issue](https://gitlab.gnome.org/GNOME/gimp/-/issues/6479) looks related.

**Comment by bionade24** [comment]
So it's a GTK bug again ? :unamused:

**Comment by mesonium** [comment]
There are also issues e.g. for [Inkscape](https://gitlab.com/inkscape/inbox/-/issues/5220)  and [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1743144) 
The [related issue](https://gitlab.gnome.org/GNOME/gtk/-/issues/4514) on the GTK bug tracker reads
> GDK calling _exit on wayland errors makes errors uncatchable by host applications.

And there's a more [in depth discussion](https://gitlab.freedesktop.org/wayland/wayland/-/issues/159#note_1074515) on the Freedesktop.org Wayland Bugtracker: 
> Locked clients that received too many events get destroyed even though the Wayland connection is not broke

 and a [potential fix](https://gitlab.freedesktop.org/wayland/wayland/-/issues/159#note_1074515):
>  Dynamically resize connection buffers 

in the pipeline, which probably tackles the underlying issue.

**Comment by mesonium** [comment]
So to me it looks like that Xournal++ fills up the input event queue without processing it fast enough every now and then, which results in `Lost connection to Wayland compositor`. 

I see this most often under openSUSE TW with Plasma 5.25.3 Wayland, when using the eraser tool in annotated PDFs using a Wacom Intuos M tablet: The changes are not applied instantly if at all, but then Xournal++'s window freezes and finally I get the error from above.

**Comment by bhennion** [comment]
@mesonium Thanks a lot for gathering this valuable information! From what you say, this is not quite only a dependency issue, but also a performance issue on our end (we need to process events faster).

> I see this most often under openSUSE TW with Plasma 5.25.3 Wayland, when using the eraser tool in annotated PDFs using a Wacom Intuos M tablet: The changes are not applied instantly if at all, but then Xournal++'s window freezes and finally I get the error from above.

Could you add some precision about your set up here (so that we know how bad things are):
1. What version of Xournal++ are you using? Is the behaviour the same, better or worse with the latest master branch?
2. What type of eraser are you using (in the app: normal, delete stroke or whiteout)? I assume normal, as this should be the most costful.
3. Do you have an estimate of the averate motion event rate of your device? (I don't know of an easy way to get that)

**Comment by mesonium** [comment]
> What version of Xournal++ are you using?

The issue happens with `1.1.1 (compiled at Feb 22 2022), libgtk 3.24.34` from the openSUSE TW repo.

>  Is the behaviour the same, better or worse with the latest master branch?
No! :) ! have compared it now with `1.1.1+dev (Git commit b5da66c9)` with the same libgtk compiled locally.
Erasing has become much smoother and I was not capable of reproducing the issue so far.

> What type of eraser are you using (in the app: normal, delete stroke or whiteout)? I assume normal, as this should be the most costful.

Yes, normal.

> Do you have an estimate of the averate motion event rate of your device? (I don't know of an easy way to get that)

I tried to estimate it using 

` sudo cat /dev/input/event23 | tee wacom`
and counting the number of events (each 24 bytes) per second in a hexeditor editor:

![Screenshot_Wacom_Events](https://user-images.githubusercontent.com/93607429/181016466-8df3b57a-375a-4491-a352-54df7b813b17.png)

With that I get an estimate of roughly `550-780` events per second.

**Comment by bhennion** [comment]
> Erasing has become much smoother and I was not capable of reproducing the issue so far.

That's good!

> and counting the number of events (each 24 bytes) per second in a hexeditor editor:

I don't think events all have the same size when they reach our end.
I think a more relyable estimation method is using libinput:
```
$ libinput record > record.yml
$ libinput analyze recording record.yml > analysis.out
```
Then counting lines and looking at timestamps in the analysis should give a decent estimate.
For an Intuos Pro M, I get ~200 events/sec. On another device (Latitude 7410 on-screen tablet), I get ~280 events/sec.
Those numbers correspond (more or less) to the number of events Xournal++ had to treat every second (I counted those on a debug version).
This most likely depends on what you do with you stylus though. My statistics come from recording while actually writing.

Anyway, we can still agregate events ourselves to reduce the performance costs if need be.

**Comment by mesonium** [comment]
> I think a more relyable estimation method is using libinput:
Yes, that makes probably more sense. That way I get ~133 events/sec using the normal easer with my Wacom Intuos M.

**Comment by ElisabettaCarrara** [comment]
Just to add my report, I am on linux manjaro Plasma KDE. Every time I open the pamac GUI it says it can not open it due to "lost connection to bus". It takes coaxing to open it from termina (not advised I know!). Today I caught that error above after coaxing it open from terminal, and it closing abruptly after finishing the install process of the software I was installing.

**Comment by Ranguvar** [comment]
Just a note that Mozilla ended up working around this with a proxy in FF 121. There's also a Rust implementation.

https://bugzilla.mozilla.org/show_bug.cgi?id=1743144
